### PR TITLE
Ensure SDK credential cache type safety

### PR DESCRIPTION
**Pull request by ysaito1001**
## Motivation and Context
Implements [RFC 28](https://github.com/awslabs/smithy-rs/blob/main/design/src/rfcs/rfc0028_sdk_credential_cache_type_safety.md). Built on top of #2108.

## Description
This PR improves APIs around credentials caching and credentials provider. The credentials cache previously named `LazyCachingCredentialsProvider` has been renamed to `LazyCredentialsCache` and no longer treated as a credentials provider. Customers do not interact with `LazyCredentialsCache` directly, and instead use  `CredentialsCache` that understands different flavors of credentials caching, which only supports `Lazy` credentials cache as of today (12/23/2022).

As a result of this change, a new builder method called `credentials_cache` has been introduced to `aws_config::ConfigLoader`, `aws_types::sdk_config::Builder`, and the builder for a service-specific `config::Builder`. This method takes an owned `CredentialsCache` as input, and the `load` method of a service-specific builder is a single-sourced place to create an actual credentials cache `LazyCredentialsCache` out of the `CredentialsCache` and the credentials provider at hand. The primary use cases affected have been documented in `CHANGELOG.next.toml`.

Now that the `load` method of a service-specific builder combines a credentials cache and a credentials provider, ensuring `LazyCredentialsCache` is on by default, no credentials provider needs to store a credentials cache explicitly as previously done by `DefaultCredentialsChain` or `AssumeRoleProvider`; those providers do not store a credentials cache anymore.

## Testing
- [x] Passed tests in CI

## Checklist

- [x] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates
- [x] I have included examples in `CHANGELOG.next.toml` showing how to use `credentials_provider` and `credentials_cache` methods on `ConfigLoader`.

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: aws/rust-runtime/aws-config/src/lib.rs
Comment (Velfi) [inline aws/rust-runtime/aws-config/src/lib.rs:497]: IMO, using `.unwrap_or_else()` is preferred to having two assignments on one line.

File: aws/rust-runtime/aws-credential-types/src/cache/lazy_caching.rs
Comment (Velfi) [inline aws/rust-runtime/aws-credential-types/src/cache/lazy_caching.rs:131]: Does the `CredentialCache` link here work? I can't remember if doc.rs handles links containing backticks correctly.

File: aws/rust-runtime/aws-http/src/auth.rs
Comment (Velfi) [inline aws/rust-runtime/aws-http/src/auth.rs:189]: I'm thinking we should consider adding a `test` method to `Credentials` with a `#[cfg(test)]` gate. The function would return these credentials or similar.

File: aws/rust-runtime/aws-types/src/sdk_config.rs
Comment (Velfi) [inline aws/rust-runtime/aws-types/src/sdk_config.rs:334]: We should consider linking here to types/docs that describe in greater depth how configuring credential caching works.

File: aws/sdk/integration-tests/s3/tests/user-agent-app-name.rs
Comment (Velfi) [inline aws/sdk/integration-tests/s3/tests/user-agent-app-name.rs:18]: This is another place where we could be calling a `Credentials::test` method to create creds.

File: aws/rust-runtime/aws-config/src/lib.rs
Comment (ysaito1001) [reply aws/rust-runtime/aws-config/src/lib.rs]: Agreed. Addressed in af61b54. Tried to convert other occurrences in the same method but they slightly differ and could not be converted, e.g. some of them have `.await` in their else blocks.

File: aws/rust-runtime/aws-credential-types/src/cache/lazy_caching.rs
Comment (ysaito1001) [reply aws/rust-runtime/aws-credential-types/src/cache/lazy_caching.rs:131]: Yes, confirmed it works.

File: aws/rust-runtime/aws-types/src/sdk_config.rs
Comment (ysaito1001) [reply aws/rust-runtime/aws-types/src/sdk_config.rs]: Addressed in 1368f00.

File: aws/rust-runtime/aws-http/src/auth.rs
Comment (ysaito1001) [reply aws/rust-runtime/aws-http/src/auth.rs:189]: That's good idea. I'll tackle it in a separate PR as it will probably affect many call sites currently creating a test `Credentials`.

File: aws/sdk/integration-tests/s3/tests/user-agent-app-name.rs
Comment (ysaito1001) [reply aws/sdk/integration-tests/s3/tests/user-agent-app-name.rs:18]: Will address it in a separate PR.

File: aws/rust-runtime/aws-http/src/auth.rs
Comment (ysaito1001) [reply aws/rust-runtime/aws-http/src/auth.rs:189]: [PR](https://github.com/awslabs/smithy-rs/pull/2145) created.

File: aws/rust-runtime/aws-config/src/sts/assume_role.rs
Comment (jdisanti) [inline aws/rust-runtime/aws-config/src/sts/assume_role.rs:195]: This cache should be configurable when creating an `AssumeRoleProvider`.

File: aws/rust-runtime/aws-http/src/auth.rs
Comment (jdisanti) [inline aws/rust-runtime/aws-http/src/auth.rs:15]: This should get called out as its own breaking change, as this is called directly by some customers. We should also make sure it's possible to directly create the `SharedCredentialsCache` without using `aws-config` (would be good to include as an example for the changelog).

**Review by jdisanti — COMMENTED** [review]
Very nice work on this! I especially like the changelog entry with all its examples. Just needs a couple minor changes before approval.

File: aws/rust-runtime/aws-config/src/sts/assume_role.rs
Comment (ysaito1001) [reply aws/rust-runtime/aws-config/src/sts/assume_role.rs]: 👍  Addressed in b987538.

File: aws/rust-runtime/aws-http/src/auth.rs
Comment (ysaito1001) [reply aws/rust-runtime/aws-http/src/auth.rs:15]: Thank you for pointing it out. That's certainly another place causing a breaking change. Provided the change log and an example in 9b4c55f.

**Review by jdisanti — APPROVED** [review]

**Pull request merged**
